src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp

```diff
@@ -442,17 +448,81 @@ REDHAWK_PALEXPORT void REDHAWK_PALAPI PalRestoreContext(CONTEXT * pCtx)
 
 static PalHijackCallback g_pHijackCallback;
 
+#ifdef FEATURE_SPECIAL_USER_MODE_APC
+typedef BOOL (WINAPI* QueueUserAPC2Proc)(PAPCFUNC ApcRoutine, HANDLE Thread, ULONG_PTR Data, QUEUE_USER_APC_FLAGS Flags);
+static QueueUserAPC2Proc pfnQueueUserAPC2Proc;
+
+static const QUEUE_USER_APC_FLAGS SpecialUserModeApcWithContextFlags = (QUEUE_USER_APC_FLAGS)
+                                    (QUEUE_USER_APC_FLAGS_SPECIAL_USER_APC |
+                                    QUEUE_USER_APC_CALLBACK_DATA_CONTEXT);
+
+static void NTAPI EmptyActivationHandler(ULONG_PTR parameter){}
+
+static void NTAPI ActivationHandler(ULONG_PTR parameter)
+{
+    APC_CALLBACK_DATA* data = (APC_CALLBACK_DATA*)parameter;
+    g_pHijackCallback(data->ContextRecord, NULL);
+
+    Thread* pThread = (Thread*)data->Parameter;
+    pThread->SetActivationPending(false);
+}
+#endif
+
 REDHAWK_PALEXPORT UInt32_BOOL REDHAWK_PALAPI PalRegisterHijackCallback(_In_ PalHijackCallback callback)
 {
     ASSERT(g_pHijackCallback == NULL);
     g_pHijackCallback = callback;
 
+#ifdef FEATURE_SPECIAL_USER_MODE_APC
+    // Check if we have QueueUserAPC2
+    HMODULE hKernel32 = LoadKernel32dll();
+    pfnQueueUserAPC2Proc = (QueueUserAPC2Proc)GetProcAddress(hKernel32, "QueueUserAPC2");
+
+    // See if QueueUserAPC2 supports the special user-mode APC with a callback that includes the interrupted CONTEXT
+    if (pfnQueueUserAPC2Proc != NULL)
+    {
+        if (!(*pfnQueueUserAPC2Proc)(EmptyActivationHandler, GetCurrentThread(), 0, SpecialUserModeApcWithContextFlags))
+        {
+            pfnQueueUserAPC2Proc = NULL;
+        }
+    }
+#endif
+
     return true;
 }
 
 REDHAWK_PALEXPORT void REDHAWK_PALAPI PalHijack(HANDLE hThread, _In_opt_ void* pThreadToHijack)
 {
     _ASSERTE(hThread != INVALID_HANDLE_VALUE);
+
+#ifdef FEATURE_SPECIAL_USER_MODE_APC
+    if (pfnQueueUserAPC2Proc)
+    {
+        Thread* pThread = (Thread*)pThreadToHijack;
+
+        // An APC can be interrupted by another one, do not queue more if one is pending.
+        if (!pThread->IsActivationPending())
+        {
+            pThread->SetActivationPending(true);
+            BOOL success = pfnQueueUserAPC2Proc(
+                &ActivationHandler,
+                hThread,
+                (ULONG_PTR)pThreadToHijack,
+                SpecialUserModeApcWithContextFlags);
+
+            if (!success)
+            {
+                // Failure to send the signal is fatal.
+                // There are no known transient reasons for this to fail.
+                ASSERT(!"failed to queue an APC");
+                abort();
```

Comment:
We seem to using `RhFailFast` in Windows PAL

Reply:
Yes, `RhFailFast` may be better.  I am also thinking maybe not crash here? I am not 100% sure there cannot be a transient failure to queue. 

I looked through what happens in this call on the OS side and it can fail, but the reasons are basically either invalid arguments or something is wrong with the target thread and it is unlikely to go away on the next try.

I am not 100% sure though. I've been hesitating a lot on this one.

Reply:
Can this fail on out-of-memory?

Reply:
Unix has a few rare scenarios where queuing a signal can fail transiently - like when a thread is about to exit, OSX may refuse to signal it. - At the moment we decide to send a signal the thread is alive and runs managed code, but when we get to sending it is already native and stopped on theadstore lock trying to leave runtime for good. This is where we get ESRCH error. On the next iteration we will see the thread is no longer running managed code.
This is a rare race, but we have seen it happening.


Perhaps something like this is possible on Windows.

Reply:
> Can this fail on out-of-memory?

Actually yes - there is a part that needs to allocate a small piece of nonpaged memory to put the APC in the list of APCs hanging off the thread before scheduling/signaling the thread. The insertion can fail with `STATUS_NO_MEMORY`.

I guess we should not fail then.

Reply:
I will keep the assert though - in case this happens in tests, which is more likely to indicate that something is broken.